### PR TITLE
Rework subscriptions to not cause re-renders

### PR DESCRIPTION
**Pull request by albertogasparin**
This is the first batch of backported changes to improve RLL internals to reduce leverage on React internals and re-render cycles. 

Notable changes:
- Now subscriptions are handled fully outside react, without needing re-renders to trigger load. 
- Wait/Phase contexts now share the same API, so subscription code can be the same
- If no SSR, we do not hydrate immediately anymore, just wait the same way as SSR does
- `afterPaint` preloads are now triggered in `requestIdleCallback` to now affect critical load
- no longer publish source maps on builds

### CODE REVIEW

File: examples/playground/components/after-paint/with-ssr.tsx
Comment (albertogasparin) [inline examples/playground/components/after-paint/with-ssr.tsx:3]: These examples changes are just quality-of-life changes, making them more strict in behaviour and fully StrictMode compatible

File: packages/core/react-loosely-lazy/__tests__/integration.test.tsx
Comment (albertogasparin) [inline packages/core/react-loosely-lazy/__tests__/integration.test.tsx:134]: Change in behaviour: in case we render the fallback on SSR, we no longer hydrate it immediately (with a live fallback) as that is not how real Suspense works. So we better align

File: packages/core/react-loosely-lazy/src/lazy-wait/__tests__/lazy-wait.integration.test.tsx
Comment (albertogasparin) [inline packages/core/react-loosely-lazy/src/lazy-wait/__tests__/lazy-wait.integration.test.tsx:25]: New implementation triggers double side effects on all types. We rely on the loader (webpack/parcel) to be smart. That's ok, right?

File: packages/core/react-loosely-lazy/src/lazy-wait/context.ts
Comment (albertogasparin) [inline packages/core/react-loosely-lazy/src/lazy-wait/context.ts:8]: Align the api to PhaseContext so the subscription logic can be shared. Also renamed as `Until*` makes it hard to mentally correlate with `LazyWait`

File: packages/core/react-loosely-lazy/src/lazy/__tests__/lazy-client.unit.test.tsx
Comment (albertogasparin) [inline packages/core/react-loosely-lazy/src/lazy/__tests__/lazy-client.unit.test.tsx:317]: This is now needed as we use `requestIdleCallback` so preloads are added on next tick

File: packages/core/react-loosely-lazy/src/lazy/components/client.tsx
Comment (albertogasparin) [inline packages/core/react-loosely-lazy/src/lazy/components/client.tsx:72]: Currently subscribe to phases only for `afterPaint` (as `lazy` already does not wait for phases, right?), but once we merge #73 this will apply for both `afterPaint`  and `lazy` I think

File: packages/core/react-loosely-lazy/src/lazy/components/client.tsx
Comment (albertogasparin) [inline packages/core/react-loosely-lazy/src/lazy/components/client.tsx:84]: Given `useEffect` is not called in "real" suspended trees, move to `requestIdleCallback` so we do not impact render / critical path

File: packages/core/react-loosely-lazy/src/lazy/deferred.ts
Comment (albertogasparin) [inline packages/core/react-loosely-lazy/src/lazy/deferred.ts:16]: Copied these changes from #73 as seems like a worthy fix

File: packages/core/react-loosely-lazy/src/lazy-wait/__tests__/lazy-wait.integration.test.tsx
Comment (MonicaOlejniczak) [reply packages/core/react-loosely-lazy/src/lazy-wait/__tests__/lazy-wait.integration.test.tsx:25]: Why does it double exactly?

File: packages/core/react-loosely-lazy/src/lazy-wait/main.tsx
Comment (albertogasparin) [reply packages/core/react-loosely-lazy/src/lazy-wait/main.tsx:15]: True, but then was having complications on the subscription side, where one returns numbers (Phase) but the other boolean (Wait). So decided to stick with numbers everywhere.

File: packages/core/react-loosely-lazy/src/lazy-wait/__tests__/lazy-wait.integration.test.tsx
Comment (albertogasparin) [reply packages/core/react-loosely-lazy/src/lazy-wait/__tests__/lazy-wait.integration.test.tsx:25]: Because `useSubscription` uses `useMemo`, so it gets triggered twice for `lazy` too (the code paths between `lazyForPaint` and `lazy` are now the same) while before `lazy` used `useEffect`.

File: packages/core/react-loosely-lazy/src/lazy/components/client.tsx
Comment (albertogasparin) [reply packages/core/react-loosely-lazy/src/lazy/components/client.tsx:66]: Hmmm... chose `noWait` as the boolean is flipped. The fact that `wait false = go` and `wait true = hold on` is the opposite than Phase (where `phase false = hold on` and `phase true = go`). 
I'd like to reconcile the booleans so it is easier to reason about here. 
Suggestions?

File: packages/core/react-loosely-lazy/src/lazy/components/utils.ts
Comment (albertogasparin) [reply packages/core/react-loosely-lazy/src/lazy/components/utils.ts:27]: 🤔 Can you provide an example implementation?

File: packages/core/react-loosely-lazy/src/lazy/types.ts
Comment (albertogasparin) [reply packages/core/react-loosely-lazy/src/lazy/types.ts:32]: It is shared between `lazy`, `lazy-wait` and `phase` (which I'd love to rename/move around as structure is a bit confusing now).

File: packages/core/react-loosely-lazy/src/lazy/components/utils.ts
Comment (MonicaOlejniczak) [reply packages/core/react-loosely-lazy/src/lazy/components/utils.ts:27]:
```
useSubscription({
      comparator: v => v === 1,
      context: WaitContext,
      load,
      onValue: (value) => status.noWait = value;
});
```

**Review by MonicaOlejniczak — APPROVED** [review]

**Pull request merged**
